src/notifications/signal.h

```diff
@@ -0,0 +1,122 @@
+/*
+ * Copyright (C) 2021 by the Widelands Development Team
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; either version 2
+ * of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
+ *
+ */
+
+#ifndef WL_NOTIFICATIONS_SIGNAL_H
+#define WL_NOTIFICATIONS_SIGNAL_H
+
+#include <cassert>
+#include <functional>
+#include <list>
+#include <memory>
+#include <set>
+
+#include "base/macros.h"
+#include "base/wexception.h"
+
+namespace Notifications {
+
+/**
+ * Whether a new subscriber will be the last (default)
+ * or first to be invoked when the signal is triggered.
+ */
+enum class SubscriberPosition { kFront, kBack };
+
+/** Class that allows attaching callback functions to an object. */
+template <typename... Args> class Signal {
+public:
+	/** Wrapper around a callback function. */
+	struct SignalSubscriber {
+		explicit SignalSubscriber(const Signal& p, const std::function<void(Args...)>& c)
+		   : parent_(p), callback_(c) {
+		}
+		~SignalSubscriber() {
+			parent_.unsubscribe(this);
+		}
+
+		const Signal& parent_;
+		const std::function<void(Args...)> callback_;
+
+		DISALLOW_COPY_AND_ASSIGN(SignalSubscriber);
+	};
+
+	/** Invoke all the signal's subscribers' callback functions. */
+	void operator()(Args... args) const {
+		for (const auto& s : all_subscribers_) {
+			s->callback_(args...);
+		}
+	}
+
+	/** Create a subscriber with a user-defined lifetime. */
+	std::unique_ptr<SignalSubscriber>
+	subscribe(const std::function<void(Args...)>& callback,
+	          SubscriberPosition pos = SubscriberPosition::kBack) const {
+		SignalSubscriber* s = new SignalSubscriber(*this, callback);
+		switch (pos) {
+		case SubscriberPosition::kBack:
+			all_subscribers_.push_back(s);
+			break;
+		case SubscriberPosition::kFront:
+			all_subscribers_.push_front(s);
+			break;
+		}
+		return std::unique_ptr<SignalSubscriber>(s);
+	}
+
+	/** Create a subscriber with the same lifetime as the signal. */
+	inline void connect(const std::function<void(Args...)>& callback,
+	                    SubscriberPosition pos = SubscriberPosition::kBack) const {
+		owned_subscribers_.insert(subscribe(callback, pos));
+	}
+
+	/** Create a subscriber that echoes the signal's invokations to another signal. */
```

Comment:
I do not understand this comment. What does echo mean? How do you avoid cycles?

Reply:
This function seems to assume that the other signal outlives this signal. This should be mentioned in the comments